### PR TITLE
Fix bug with ComponentResource resource refs

### PR DESCRIPTION
**Pull request by lblackstone**
Now that resources are serialized as refs, ComponentResources
may try to unmarshal local resource refs before they are
initialized during the RegisterResource step.

This change avoids that issue by skipping Output marshaling
for local ComponentResources during the RegisterResource step.
These Outputs will be handled instead during the
RegisterResourceOutputs step.

Fix https://github.com/pulumi/pulumi/issues/5940

### CODE REVIEW

File: pkg/resource/deploy/source_eval.go [hunk 1]
```diff
@@ -929,6 +929,10 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 
 	// Filter out partially-known values if the requestor does not support them.
 	outputs := result.State.Outputs
+	if !req.GetCustom() && !remote {
+		// Local ComponentResources may contain unresolved resource refs, so ignore those outputs.
+		outputs = resource.PropertyMap{}
```
Comment (pgavlin) [inline pkg/resource/deploy/source_eval.go:934]: It would be good to add a note about the root cause of the unresolvable refs (namely, the code in `SameStep.Apply`) and how we would fix the underlying issue. Could we open an issue to track that fix?

Comment (lblackstone) [reply pkg/resource/deploy/source_eval.go:934]: @pgavlin Would you mind writing that note? I'm not sure fully grasp the subtleties here.

Comment (lblackstone) [reply pkg/resource/deploy/source_eval.go:934]: Actually, let me summarize my understanding, and you can correct me if I'm wrong:

In the case of a no-op preview step, the old input state is copied to the new input state. As a result,
any resource refs are marshaled/unmarshaled in the SDK code. ComponentResources are not guaranteed
to register resources until the RegisterOutputResources method has been called, which leads to a race
condition where the SDK invokes the `getResource` method to resolve a resource reference prior to
the resource being ready/registered.

Comment (pgavlin) [reply pkg/resource/deploy/source_eval.go:934]: > In the case of a no-op preview step, the old input state is copied to the new input state. As a result,
any resource refs are marshaled/unmarshaled in the SDK code. ComponentResources are not guaranteed
to register resources until the RegisterOutputResources method has been called, which leads to a race
condition where the SDK invokes the getResource method to resolve a resource reference prior to
the resource being ready/registered.

I'd rephrase as follows:

"In the case of a `SameStep`, the old resource outputs are returned to the language host after the step is executed. The outputs of a `ComponentResource` may depend on resources that have not been registered at the time the `ComponentResource` is itself registered, as the outputs are set by a later call to `RegisterResourceOutputs`. Therefore, when the `SameStep` returns the old resource outputs for a `ComponentResource`, it may return references to resources that have not yet been registered, which will cause the SDK's calls to `getResource` to fail when it attempts to resolve those references.

> One potential fix would be to change the behavior of pulumi:pulumi:getResource to block until a register resource for the URN comes in (failing if the deployment completes without the URN being seen).

This seems dangerous, as it could trivially lead to deadlock (i.e. the program is waiting for a resource that is never registered, thus keeping the deployment from ever completing). I think that the right long-term fix is to change the behavior of `SameStep` to only return old outputs for custom resources. In theory, the language SDK shouldn't even need the old outputs for a component resource--it simply deserializes them and discards them.

**Review by pgavlin — APPROVED** [review]

**Comment by pgavlin** [comment]
Any way we could get a simple unit test together?
